nservicebus/messaging/uniformsession.md

```diff
@@ -0,0 +1,50 @@
+summary: Uniform Session allows to bridge the gap between the message session and the pipeline context and enables dependency injection.
+reviewed: 2017-10-27
+component: UniformSession
+---
+
+NServiceBus v6 and higher introduced a few design changes that directly impact all the code that is using previous versions of NServiceBus. The [changes](/nservicebus/upgrades/5to6/moving-away-from-ibus.md) are:
+
+- In previous versions, the `IBus` interface was automatically registered in the IOC container. In Version 6, the new context-aware interfaces, for example, `IEndpointInstance`, `IMessageSession` and `IMessageHandlerContext`, etc., are not automatically registered in dependency injection.
+- In Versions 5 and below, when a custom component was registered in the container, the custom component had access to the `IBus` instance via dependency injection.
+- Clear separation of concerns between message operations outside the message handling pipeline (`IEndpointInstance` and `IMessageSession`) vs. inside the message handling pipeline (`IMessageHandlerContext`). Operations inside the message handling pipeline enlist in the transaction available and participate in the batching operations of the transport. Operations outside the pipeline don't have those capabilities. The ambivalent behavior of the `IBus` interface is therefore replaced with explicit and intention revealing interfaces.
+- Message operations are asynchronous by default
+
+These design changes enable a path to a better design because:
+
+- Child containers used no longer need to be mutated when a message arrives to rebind the current message handler context which enables wider container integration especially for immutable containers
+- If a component needs state that is created by the framework per message handling pipeline invocation the state has to be either
+    - Explicitly carried into the components that need it and thus the component callstack will need to become asynchronous or
+    - Managed in the ports and adapters of the system, i.ex. the message handler which is already asynchronous thus the component hierarchy can remain synchronous because the side effect incurring I/O operations are no longer the domain logic's concern.
+- The domain logic becomes side-effect free and no longer produces messages but returns an explicit decision matrix which will be interpreted by the caller to create the necessary side effects
+
+It is advised to embrace the design approach if possible. That being said for customers transitioning from the previous version to NServiceBus v6 and higher the design decision caused some grief. This package reintroduces an opt-in approach for a uniform session approach that works seamlessly as a message session outside the pipeline and as a pipeline context inside the message handling pipeline. The message operations provided on the uniform session represent a common denominator but do not support more advanced scenarios like persistence session access.
```

Comment:
from the previous version to NServiceBus v6 and higher


>from previous versions

since we already stated that this is <v6 above